### PR TITLE
#165591444 Refactor API endpoints

### PR DESCRIPTION
**Pull request by chokonaira**
What does this PR do?
Implement feedback

Description of Task to be completed?

Refactored all controllers and added models
Updated READMe file
Fixed code climate test coverage reporter
Dist should build inside app dir not outside
Remove PDF document
Account type should only accept two account types
Remove status from req.body in user create account
Fixed all validation errors
Remove dead codes and eslint rules
modify status code
Refactored all test

have the following endpoint working
GET /api/v1/admin/email/accounts

How should this be manually tested?
https://github.com/chokonaira/Banka.git

on your terminal run cd banka

run git checkout ft-refactor-API-endpoints-165591444

run npm install

run npm run before:start
run npm run dev

make a request to all endpoints stated in READMe

Any background context you want to provide?
NA
What are the relevant pivotal tracker stories?
#165591444

### CODE REVIEW

**Comment by slimsolz** [comment]
@chokonaira in your description, under the question What does this PR do?
you could just say `Implement feedback`

File: app/controllers/adminController.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        error: 'Unable to get all account details!! Server Error, Please Try Again',
+      });
+    }
+  }
+
+  static async deleteAccount(req, res) {
+    try {
+      const isDeleted = await accountModel.delete(req, res);
+      if (isDeleted > 0) {
+        return res.status(200).send({
```
Comment (slimsolz) [inline app/controllers/adminController.js:67]: Is there a reason you have 200 on this line but have 204 on the next?

File: app/controllers/cashierController.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      const createdOn = new Date();
+      const lastTransaction = await transactionModel.getLastTransaction(req);
+      if (lastTransaction) {
+        oldBalance = lastTransaction.accountbalance;
+      } else {
+        const bankAccount = await accountModel.getAccount(req, res);
+        oldBalance = bankAccount[0].openingbalance;
+      }
+      if ((+oldBalance) < (+amount)) {
+        return res.status(409).json({
```
Comment (slimsolz) [inline app/controllers/cashierController.js]: 409 is code for conflict. I think 400 is better for this

File: app/models/accountModel.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  }
+
+  static async getAccount(req, res) {
+    const query = 'SELECT * FROM accounts WHERE accountNo = $1';
+    const values = [req.params.accountNumber];
+    try {
+      const { rows } = await pool.query(query, values);
+      if (!rows[0]) {
+        return res.status(200).send({
```
Comment (slimsolz) [inline app/models/accountModel.js]: 200 on this line and 204 on the next. fix this

File: app/controllers/adminController.js (hunk 1 above)
Comment (chokonaira) [reply app/controllers/adminController.js:67]: not really, an oversight, will fix it

File: app/models/accountModel.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  }
+
+  static async getAccountByEmail(req, res) {
+    const allAccountsQuery = 'SELECT * FROM accounts WHERE owner = (SELECT user_id FROM users WHERE email = $1)';
+    const values = [req.params.userEmail];
+    try {
+      const { rows } = await pool.query(allAccountsQuery, values);
+      if (!rows[0]) {
+        return res.status(200).send({
```
Comment (slimsolz) [inline app/models/accountModel.js]: fix this also

File: app/models/accountModel.js (hunk 3 above)
Comment (chokonaira) [reply app/models/accountModel.js]: okay will fix it

File: app/models/accountModel.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    if (req.query.status) {
+      allAccountsQuery = 'SELECT * FROM accounts WHERE status = $1 ORDER BY createdOn DESC';
+      values = [req.query.status];
+    } else {
+      allAccountsQuery = 'SELECT * FROM accounts ORDER BY createdOn DESC';
+    }
+    try {
+      const { rows } = await pool.query(allAccountsQuery, values);
+      if (!rows[0]) {
+        return res.status(200).send({
```
Comment (slimsolz) [inline app/models/accountModel.js]: this too

File: app/models/accountModel.js (hunk 4 above)
Comment (chokonaira) [reply app/models/accountModel.js]: okay

File: app/models/accountModel.js (hunk 5 above)
Comment (chokonaira) [reply app/models/accountModel.js]: okay

File: app/controllers/cashierController.js (hunk 2 above)
Comment (chokonaira) [reply app/controllers/cashierController.js]: ouch, will fix it

File: app/models/transactionModel.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  }
+
+  static async getOneTransaction(req, res) {
+    const query = 'SELECT * FROM transactions WHERE transaction_id = $1';
+    const values = [req.params.transactionId];
+    try {
+      const { rows } = await pool.query(query, values);
+      if (rows.length === 0) {
+        return res.status(200).send({
```
Comment (slimsolz) [inline app/models/transactionModel.js]: and this

Comment (chokonaira) [reply app/models/transactionModel.js]: okay

File: app/models/userModel.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    const { firstname, lastname, email, type } = req.body;
+    const values = [firstname.trim(), lastname.trim(), email.trim(), type.trim(), hashedPassword];
+    try {
+      const { rows } = await pool.query(newUserQuery, values);
+      return rows;
+    } catch (error) {
+      if (error.routine === '_bt_check_unique') {
+        return res.status(400).send({
+          status: 400,
+          error: `User with '${req.body.email}' already exists`,
```
Comment (slimsolz) [inline app/models/userModel.js]: change this message to `user already exists` no need to include the email

Comment (chokonaira) [reply app/models/userModel.js]: okay

File: app/server.js [hunk 8]
```diff
@@ -11,12 +13,18 @@ app.use(bodyParser.json());
...
 app.get('/', (req, res) => {
   res.status(200).send(
-    'Banka is a light-weight core banking application that powers banking operations like account creation, customer deposit and withdrawals. This app is meant to support a single bank, where users can signup and create bank accounts online, but must visit the branch to withdraw or deposit money..',
+    'Welcome to Banka...',
   );
 });
+app.get('/*', (req, res) => {
+  res.status(404).send({
+    status: 404,
+    error: 'bad request, invalid route',
```
Comment (slimsolz) [inline app/server.js]: change this to `Not Found`

Comment (chokonaira) [reply app/server.js]: okay

File: app/test/unitTests/admin.js [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+chai.use(chaiHttp);
+chai.should();
+
+let adminBearerToken;
+let cashierBearerToken;
+
+describe('Admin', function () {
+  this.timeout(0);
+  before(async () => {
+    // login admin
```
Comment (slimsolz) [inline app/test/unitTests/admin.js]: please remove all the comments

File: app/test/unitTests/admin.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        .send({
+          status: 'notactive',
+        })
+        .set('authorization', adminBearerToken);
+      res.should.have.status(400);
+      res.body.should.have.property('error');
+      res.body.error.should.equal('Invalid account status field, status should be "dormant" or "active"');
+    });
+
+    it('Should return account does not exist with status 204', async () => {
```
Comment (slimsolz) [inline app/test/unitTests/admin.js]: this says should return `204` but I see you expect `200` returned. fix this

File: app/test/unitTests/admin.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        .send({
+          status: 'dormant',
+        })
+        .set('authorization', adminBearerToken);
+      res.should.have.status(200);
+      res.body.should.have.property('data');
+      res.body.data.should.have.length(0);
+    });
+
+    it('Should return account updated successfully with account data and status 201', async () => {
```
Comment (slimsolz) [inline app/test/unitTests/admin.js]: change this to 200 not 201

File: app/test/unitTests/admin.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        .request(app)
+        .get('/api/v1/accounts?status=active')
+        .set('authorization', adminBearerToken);
+      res.should.have.status(200);
+      res.body.should.have.property('data');
+    });
+  });
+
+  describe('get a specific user accounts', () => {
+    it('Should return no account found with status 204', async () => {
```
Comment (slimsolz) [inline app/test/unitTests/admin.js]: fix this too

**Review by slimsolz — CHANGES_REQUESTED** [review]
Attend to the comments

**Comment by slimsolz** [comment]
fix failing test

**Review by slimsolz — APPROVED** [review]

**Pull request merged**